### PR TITLE
 feat(datepickers): introduce DatepickerRange component

### PR DESCRIPTION
**Pull request by austin94**
## Description

This PR introduces the `DatepickerRange` component. To allow for additional flexibility in input and calendar placement, this component uses the following API:

```jsx
<DatepickerRange>
  <DatepickerRange.Start>
    <input />
  </DatepickerRange.Start>
  <DatepickerRange.End>
    <input />
  </DatepickerRange.End>
  <DatepickerRange.Calendar />
</DatepickerRange>
```

This type of composition allows consumers to determine which type of inputs will be used and how they can be laid out. In the provided example I use our `react-forms` and `react-grid` components for the dynamic layout.

## Checklist

- [x] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [ ] ~:nail_care: view component styling is based on a Garden CSS
      component~
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

File: packages/datepickers/src/styled/styled-datepicker.tsx [hunk 1]
```diff
@@ -104,10 +107,25 @@ export const StyledCalendar = styled.div<{ isSmall?: boolean }>`
...
 
+export const StyledRangeCalendar = styled.div`
+  display: flex;
+  overflow: auto;
+
+  ${StyledDatepicker} {
+    padding: 0;
+  }
+
+
```
Comment (jzempel) [inline packages/datepickers/src/styled/styled-datepicker.tsx]: delete extra space

**Comment by austin94** [comment]
> The interaction between the ranged date picker and the associated start/end inputs feels off to me

@ginnywood and I sat down and re-worked some of the focus strategy for this interaction. The component now alternates selection based on previous selection. It now also allows `ENTER` key presses to validate the date input.

**Comment by ryanseddon** [comment]
I noticed a weird bug while testing the custom format example, but it turns out you're just missing the `locale` being set for the `dateFormatter` in the `custom-formats.md` example. 

It happens for me as it tries to switch to my locale on close which is dd/mm/yyyy but the example sets the initial `locale` to `en-US` of `mm/dd/yyyy`.

datepicker

File: packages/datepickers/src/Datepicker/Datepicker.tsx [hunk 2]
```diff
@@ -328,7 +328,8 @@ Datepicker.defaultProps = {
   refKey: 'ref',
   animate: true,
   eventsEnabled: true,
-  zIndex: 1000
+  zIndex: 1000,
+  locale: 'en-US'
```
Comment (ryanseddon) [inline packages/datepickers/src/Datepicker/Datepicker.tsx]: Should we have a default? Like the issue mentioned in the PR if a user forgets to pass the locale in a custom formatter to match this default it'll have some confusing UX if they don't use the US format.

File: packages/datepickers/src/DatepickerRange/DatepickerRange.tsx [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  minValue: PropTypes.instanceOf(Date),
+  maxValue: PropTypes.instanceOf(Date),
+  onChange: PropTypes.func,
+  formatDate: PropTypes.func,
+  customParseDate: PropTypes.func,
+  small: PropTypes.bool
+};
+
+DatepickerRange.defaultProps = {
+  locale: 'en-US',
```
Comment (ryanseddon) [inline packages/datepickers/src/DatepickerRange/DatepickerRange.tsx]: Same issue as above, I don't think we should have an explicit default

File: packages/datepickers/src/DatepickerRange/components/Calendar.tsx [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    startValue,
+    endValue,
+    onChange
+  } = useDatepickerRangeContext();
+
+  return (
+    <StyledRangeCalendar
+      data-garden-id="datepickers.range"
+      data-garden-version={PACKAGE_VERSION}
+      data-test-id="range-calendar"
```
Comment (ryanseddon) [inline packages/datepickers/src/DatepickerRange/components/Calendar.tsx]: Do we ship these test-ids to our published packages?

File: packages/datepickers/src/DatepickerRange/components/Month.tsx [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+            if (!isSelected) {
+              dispatch({ type: 'HOVER_DATE', value: date });
+            }
+          }}
+          data-test-id="day"
+          data-test-previous={isPreviousMonth}
+          data-test-selected={!isInvalidDateRange && isSelected}
+          data-test-disabled={isDisabled}
+          data-test-today={isCurrentDate}
+          data-test-hidden="false"
```
Comment (ryanseddon) [inline packages/datepickers/src/DatepickerRange/components/Month.tsx]: What's the purpose of all of these for testing only?

File: packages/datepickers/src/DatepickerRange/utils/datepicker-range-reducer.ts [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+
+      return {
+        ...state,
+        endInputValue,
+        hoverDate: undefined,
+        previewDate
+      };
+    }
+    case 'CLICK_DATE':
+      // console.log(state.isStartFocused, state.isEndFocused);
```
Comment (ryanseddon) [inline packages/datepickers/src/DatepickerRange/utils/datepicker-range-reducer.ts]: delete

File: packages/datepickers/src/DatepickerRange/components/Calendar.tsx (hunk 4 above)
Comment (austin94) [reply packages/datepickers/src/DatepickerRange/components/Calendar.tsx]: https://github.com/zendeskgarden/react-components/blob/master/babel.config.js#L26 removes all `data-test*` attributes from the `production` bundle. This should only be visible locally and in our tests.

File: packages/datepickers/src/DatepickerRange/components/Month.tsx (hunk 5 above)
Comment (austin94) [reply packages/datepickers/src/DatepickerRange/components/Month.tsx]: With us no longer using `enzyme` we aren't able to assert against props. This allows us to ensure the states are being applied without having to assert against a specific styled-components style.  This is the same process as the `Datepicker` PR.

File: packages/datepickers/src/DatepickerRange/utils/datepicker-range-reducer.ts (hunk 6 above)
Comment (austin94) [reply packages/datepickers/src/DatepickerRange/utils/datepicker-range-reducer.ts]: removed

File: packages/datepickers/src/Datepicker/Datepicker.tsx (hunk 2 above)
Comment (austin94) [reply packages/datepickers/src/Datepicker/Datepicker.tsx]: This originally came up in the `#garden` channel. In some specific locale and browser combinations, `Intl.DateTimeFormat` will use the system locale by default.

This works for formatting, but browsers are only able to natively parse in short, medium, and long `en-US` locales. (There are some slight variations between browsers, but this is mostly consistent).

Without the default we would have the Datepicker showing in `danish` for example, but unable to parse anything until MomentJS or date-fns is provided. By locking it to `en-US` it ensures that the date is always able to be parsed.

**Pull request merged**